File: JAVA/lib/pkce.java
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;

public class PKCE {
    public static String[] generate(String method) throws Exception {
        SecureRandom secureRandom = new SecureRandom();
        byte[] codeVerifierBytes = new byte[64];
        secureRandom.nextBytes(codeVerifierBytes);
        String codeVerifier = Base64.getUrlEncoder().withoutPadding().encodeToString(codeVerifierBytes);

        String codeChallenge;
        if (method.equalsIgnoreCase("plain")) {
            codeChallenge = codeVerifier;
        } else {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashed = digest.digest(codeVerifier.getBytes("US-ASCII"));
            codeChallenge = Base64.getUrlEncoder().withoutPadding().encodeToString(hashed);
        }

        return new String[]{codeVerifier, codeChallenge, method};
    }
}

File: python/lib/pkce.py
import base64
import hashlib
import os

class PKCE:
    @staticmethod
    def generate(method="S256"):
        code_verifier = base64.urlsafe_b64encode(os.urandom(64)).rstrip(b'=').decode('utf-8')

        if method.upper() == "PLAIN":
            code_challenge = code_verifier
        else:
            digest = hashlib.sha256(code_verifier.encode("ascii")).digest()
            code_challenge = base64.urlsafe_b64encode(digest).rstrip(b'=').decode('utf-8')

        return code_verifier, code_challenge, method
